enterprise/frontend/src/metabase-enterprise/application_permissions/selectors.ts

```diff
@@ -33,7 +33,7 @@ export function getPermissionWarning(
 }
 
 export const canManageSubscriptions = (state: ApplicationPermissionsState) =>
-  state.currentUser.permissions?.can_access_subscription ?? false;
+  state.currentUser?.permissions?.can_access_subscription ?? false;
```

Comment:
When using the actions/selectors without passing them down the chain, we may or may not be sure that the application has loaded. So we might not actually know if the `currentUser` is loaded yet (and I ran into a bunch of `currentUser is undefined ` errors too)

Reply:
But what's the difference between this version and the version before that it didn't require optional chaining here?

This breaks the type since our type already says `currentUser` couldn't be optional.

Does that mean our type has been wrong all the time?

Reply:
let me look into it a little more so i can give you a better explanation

Reply:
Ok I've been trying to replicate what I have been trying to mention here, but for some reason I can't anymore. So I've reverted this change

Reply:
I managed to find the issue again - it's actually a problem in the E2E environment where it tries to load the current user, presumably before the whole store has loaded (since in a real life scenario, the store + current user are already loaded). The culprit is the `canManageSubscriptions` selector just not being able to get the `currentUser`

It turns out that when we're retrieving the `currentUser` it can be typed as `User | null` . However, the `currentUser` for `ApplicationPermissionsState` is actually `UserWithApplicationPermissions`. So what I'll do is:

*   Change the type of `currentUser` in `ApplicationPermissionsState` to `UserWithApplicationPermissions | null`
*   Convert `canManageSubscriptions` to a `createSelector` that uses `(state: ApplicationPermissionsState) => state.currentUser` to select the user.
*   Add `can_manage_subscriptions?: boolean;` to `UserWithApplicationPermissions` as it was missing.
*   Add optional chaining to replicate the same behavior that we have for other parts of the app

So we'll be left with:

```
export interface ApplicationPermissionsState extends State {
  currentUser: UserWithApplicationPermissions | null;
  ...
}


export const canManageSubscriptions = createSelector(
  (state: ApplicationPermissionsState) => state.currentUser,
  user => user?.permissions?.can_manage_subscriptions ?? false,
);
```